deploy/runtools/user_topology.py

```diff
@@ -261,29 +273,32 @@ def supernode_flatten(arr):
             res = res + x
         return res
 
-    def supernode_example_6config(self):
+    def supernode_example_6config(self) -> None:
         self.roots = [FireSimSwitchNode()]
-        servers = [FireSimSuperNodeServerNode()] + [FireSimDummyServerNode() for x in range(5)]
```

Comment:
This errored since a `List[SuperNode]` couldn't be added to a `List[DummyNode]`s.

Reply:
Even if you do this?
```
servers : List[FireSimServerNode] = [FireSimSuperNodeServerNode()] + [FireSimDummyServerNode() for x in range(5)]
```

from https://mypy.readthedocs.io/en/latest/common_issues.html#variance

Reply:
For my own edification, I confirmed that with this:
```
diff --git a/deploy/runtools/user_topology.py b/deploy/runtools/user_topology.py
index 4336b4b14..a7173fd37 100644
--- a/deploy/runtools/user_topology.py
+++ b/deploy/runtools/user_topology.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 
 from runtools.firesim_topology_elements import FireSimSwitchNode, FireSimServerNode, FireSimSuperNodeServerNode, FireSimDummyServerNode, FireSimNode
 
-from typing import Optional, Union, Callable, Sequence, TYPE_CHECKING
+from typing import List, Optional, Union, Callable, Sequence, TYPE_CHECKING
 if TYPE_CHECKING:
     from runtools.firesim_topology_with_passes import FireSimTopologyWithPasses
 
@@ -275,8 +275,8 @@ class UserTopologies:
 
     def supernode_example_6config(self) -> None:
         self.roots = [FireSimSwitchNode()]
-        self.roots[0].add_downlinks([FireSimSuperNodeServerNode()])
-        self.roots[0].add_downlinks([FireSimDummyServerNode() for x in range(5)])
+        servers : List[FireSimServerNode] = [FireSimSuperNodeServerNode()] + [FireSimDummyServerNode() for x in range(5)]
+        self.roots[0].add_downlinks(servers)
 
     def supernode_example_4config(self) -> None:
         self.roots = [FireSimSwitchNode()]
Submodule target-design/chipyard contains modified content
```

I still see:
```
\u276f ./scripts/run-py-typecheck.sh 
deploy/runtools/user_topology.py:278: error: Incompatible types in assignment (expression has type "List[FireSimSuperNodeServerNode]", variable has type "List[FireSimServerNode]")
deploy/runtools/user_topology.py:278: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
deploy/runtools/user_topology.py:278: note: Consider using "Sequence" instead, which is covariant
deploy/runtools/user_topology.py:278: error: List comprehension has incompatible type List[FireSimDummyServerNode]; expected List[FireSimSuperNodeServerNode]
Found 2 errors in 1 file (checked 22 source files)

```

Reply:
This works:
```
diff --git a/deploy/runtools/user_topology.py b/deploy/runtools/user_topology.py
index 4336b4b14..ce3479c1b 100644
--- a/deploy/runtools/user_topology.py
+++ b/deploy/runtools/user_topology.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 
 from runtools.firesim_topology_elements import FireSimSwitchNode, FireSimServerNode, FireSimSuperNodeServerNode, FireSimDummyServerNode, FireSimNode
 
-from typing import Optional, Union, Callable, Sequence, TYPE_CHECKING
+from typing import cast, List, Optional, Union, Callable, Sequence, TYPE_CHECKING
 if TYPE_CHECKING:
     from runtools.firesim_topology_with_passes import FireSimTopologyWithPasses
 
@@ -275,8 +275,8 @@ class UserTopologies:
 
     def supernode_example_6config(self) -> None:
         self.roots = [FireSimSwitchNode()]
-        self.roots[0].add_downlinks([FireSimSuperNodeServerNode()])
-        self.roots[0].add_downlinks([FireSimDummyServerNode() for x in range(5)])
+        servers = [cast(FireSimServerNode, FireSimSuperNodeServerNode())] + [cast(FireSimServerNode, FireSimDummyServerNode()) for x in range(5)]
+        self.roots[0].add_downlinks(servers)
 
     def supernode_example_4config(self) -> None:
         self.roots = [FireSimSwitchNode()]
Submodule target-design/chipyard contains modified content

```

But I like the way you rewrote it to avoid the problem better.  I really hate having to rewrite reasonable code for static checkers but whatever.